pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,339 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"strings"
+
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/util/sets"
+)
+
+func mergeVirtualServicesIfNeeded(vServices []Config) (out []Config) {
+	out = make([]Config, 0, len(vServices))
+	delegatesMap := map[string]Config{}
+	// root virtualservices with delegate
+	var rootVses []Config
+
+	// 1. classify virtualservices
+	for _, vs := range vServices {
+		rule := vs.Spec.(*networking.VirtualService)
+		// it is delegate, add it to the indexer cache
+		if len(rule.Hosts) == 0 {
```

Comment:
the api will look very bad on a virtual service without hosts. Do we really really need to follow k8s and make our API look bad just to align with k8s? It feels very hacky @howardjohn such that our users wont appreciate cryptic configs even if it aligns with some k8s thing.

I still think having a proper delegate match at http level maintains a better balance between readability and flexibility. It will not have some of the ambiguities mentioned below

Reply:
>the api will look very bad on a virtual service without hosts. 

If this is controversial, i can add a new field to indicate it.

>I still think having a proper delegate match at http level maintains a better balance between readability and flexibility. It will not have some of the ambiguities mentioned below

I am not sure what this, but whatever the delegate mut be a subset of root, otherwise there is potential conflicts

Reply:
I am talking about doing the delegate outside the route block. See my proposal in the api PR. I proposed that model to avoid the complexity that is happening here.

Reply:
@howardjohn 

>SHRIRAM: Separate Include Block in VirtualService
Virtual service can have a separate include section that specifies names of other virtual services and conditions under which these virtual services should be included.
```

Eg
host: foo.com
 IncludeVirtualServices:
 - resource: ns1/vs1
    condition:
     - httpMatch:
         urimatchHasPrefix: /barbar
         hasHeaderMatchesFor: x-envoy-blah
tlsMatch:
   usesSni: foo.example.com

```

This will cause pilot to only include the http rules that have /barbar as a prefix or exact match in the vs1 virtual service in the ns1 namespace.

Reply:
Now here you can use the new `Delegate` field to check if it `Delegate` or not

Reply:
If there is no `Delegate`, we can't say it is a delegate, it can be a single standalone vs

Reply:
As per offline conv, please ignore this comment. I misunderstood the intent here.